src/MongoDB.Driver.Core/Core/Connections/ConnectionInitializer.cs

```diff
@@ -53,6 +53,14 @@ public ConnectionDescription InitializeConnection(IConnection connection, Cancel
 
             AuthenticationHelper.Authenticate(connection, description, cancellationToken);
 
+            var connectionIdServerValue = isMasterResult.ConnectionIdServerValue;
+            if (connectionIdServerValue.HasValue)
+            {
+                description = UpdateConnectionIdWithServerValue(description, connectionIdServerValue.Value);
+
+                return description;
```

Comment:
Rather than having a random `return` thrown in here in the middle of the method consider:

    if (connectionIdServerValue.HashValue)
    {
        description = UpdateConnectionIdWithServerValue(description, connectionIdServerValue.Value);
    }
    else
    {
        // old code (lines 64-74) to get connectionId use GLE
    }

Reply:
Done.